File: Shadow samurei/rust/kern/src/passwort.rs
//! Passwort-Hashes im Format von Werkzeug.
//!
//! Damit Weboberfläche (Rust) und Desktop-Oberfläche (Python) dieselbe
//! Datei `data/users.json` nutzen können, wird genau das Format von
//! `werkzeug.security` erzeugt und geprüft:
//!
//! * `scrypt:n:r:p$salz$hex` – Standard neuerer Werkzeug-Fassungen,
//! * `pbkdf2:sha256:runden$salz$hex` – Format älterer Fassungen.
//!
//! Der Vergleich der Hashes läuft zeitunabhängig (`subtle`).

use rand::Rng;
use sha2::Sha256;
use subtle::ConstantTimeEq;

/// Länge des abgeleiteten Schlüssels bei `scrypt` (wie bei Werkzeug).
const SCRYPT_LAENGE: usize = 64;
/// Standardparameter von Werkzeug für `scrypt`.
const SCRYPT_N: u32 = 32768;
const SCRYPT_R: u32 = 8;
const SCRYPT_P: u32 = 1;
/// Länge des Salzes in Zeichen (wie bei Werkzeug).
const SALZ_LAENGE: usize = 16;
/// Zeichenvorrat des Salzes (wie bei Werkzeug: Buchstaben und Ziffern).
const SALZ_ZEICHEN: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

/// Erzeugt ein zufälliges Salz.
fn salz() -> String {
    let mut zufall = rand::thread_rng();
    (0..SALZ_LAENGE)
        .map(|_| SALZ_ZEICHEN[zufall.gen_range(0..SALZ_ZEICHEN.len())] as char)
        .collect()
}

/// Berechnet den `scrypt`-Hash eines Passworts.
fn scrypt_hash(passwort: &str, salz: &str, n: u32, r: u32, p: u32) -> Option<String> {
    let log_n = (n as f64).log2().round() as u8;
    if 2u64.pow(log_n as u32) != n as u64 {
        return None;
    }
    let parameter = scrypt::Params::new(log_n, r, p, SCRYPT_LAENGE).ok()?;
    let mut ausgabe = vec![0u8; SCRYPT_LAENGE];
    scrypt::scrypt(passwort.as_bytes(), salz.as_bytes(), &parameter, &mut ausgabe).ok()?;
    Some(hex::encode(ausgabe))
}

/// Berechnet den `pbkdf2`-Hash eines Passworts (SHA-256).
fn pbkdf2_hash(passwort: &str, salz: &str, runden: u32) -> String {
    let mut ausgabe = [0u8; 32];
    pbkdf2::pbkdf2_hmac::<Sha256>(passwort.as_bytes(), salz.as_bytes(), runden, &mut ausgabe);
    hex::encode(ausgabe)
}

/// Erzeugt einen Passwort-Hash im Werkzeug-Format (`scrypt`).
pub fn erzeuge_hash(passwort: &str) -> String {
    let salz = salz();
    match scrypt_hash(passwort, &salz, SCRYPT_N, SCRYPT_R, SCRYPT_P) {
        Some(hash) => format!("scrypt:{SCRYPT_N}:{SCRYPT_R}:{SCRYPT_P}${salz}${hash}"),
        // Sollte nicht vorkommen; dann gilt das ältere, ebenfalls
        // unterstützte Verfahren.
        None => {
            let runden = 1_000_000;
            format!(
                "pbkdf2:sha256:{runden}${salz}${}",
                pbkdf2_hash(passwort, &salz, runden)
            )
        }
    }
}

/// Prüft ein Passwort gegen einen gespeicherten Hash.
///
/// Unbekannte Verfahren und defekte Hashes ergeben `false`.
pub fn pruefe(hash: &str, passwort: &str) -> bool {
    let mut teile = hash.splitn(3, '$');
    let (Some(verfahren), Some(salz), Some(erwartet)) =
        (teile.next(), teile.next(), teile.next())
    else {
        return false;
    };
    let angaben: Vec<&str> = verfahren.split(':').collect();
    let berechnet = match angaben.as_slice() {
        ["scrypt", n, r, p] => {
            let (Ok(n), Ok(r), Ok(p)) = (n.parse::<u32>(), r.parse::<u32>(), p.parse::<u32>())
            else {
                return false;
            };
            match scrypt_hash(passwort, salz, n, r, p) {
                Some(wert) => wert,
                None => return false,
            }
        }
        ["pbkdf2", "sha256", runden] => match runden.parse::<u32>() {
            Ok(runden) => pbkdf2_hash(passwort, salz, runden),
            Err(_) => return false,
        },
        ["pbkdf2", "sha256"] => pbkdf2_hash(passwort, salz, 260_000),
        _ => return false,
    };
    berechnet.as_bytes().ct_eq(erwartet.as_bytes()).into()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn eigener_hash_wird_erkannt() {
        let hash = erzeuge_hash("geheim123");
        assert!(hash.starts_with("scrypt:32768:8:1$"));
        assert!(pruefe(&hash, "geheim123"));
        assert!(!pruefe(&hash, "falsch"));
    }

    #[test]
    fn hash_von_werkzeug_wird_geprueft() {
        // Von „werkzeug.security.generate_password_hash('1234')“ erzeugt.
        let scrypt_hash = "scrypt:32768:8:1$1x3pQmIuOUvtdKFi$65eed30c6df3fdf1abd2bfab23dc634019e91c44d1431ee33bbc6a84fd6bdae232d53f609d36d421543ebea99e2475214cdf5cf3abc04e18a35bb84f9807b22e";
        assert!(pruefe(scrypt_hash, "1234"));
        assert!(!pruefe(scrypt_hash, "12345"));
        // Format älterer Werkzeug-Fassungen.
        let pbkdf2_hash = "pbkdf2:sha256:1000000$GX5byfqQyta97TXC$1678024184e986fb24d9ab787f56c48b3ac2b5fbeeb1e7d9ed3c85125306a4d6";
        assert!(pruefe(pbkdf2_hash, "1234"));
        assert!(!pruefe(pbkdf2_hash, "falsch"));
    }

    #[test]
    fn defekte_hashes_ergeben_false() {
        assert!(!pruefe("", "1234"));
        assert!(!pruefe("ohne-trennzeichen", "1234"));
        assert!(!pruefe("md5$salz$abc", "1234"));
        assert!(!pruefe("scrypt:abc:8:1$salz$abc", "1234"));
        assert!(!pruefe("scrypt:32769:8:1$salz$abc", "1234"));
    }

    #[test]
    fn jedes_salz_ist_neu() {
        assert_ne!(erzeuge_hash("gleich"), erzeuge_hash("gleich"));
    }
}

File: Shadow samurei py/kern/passwort.py
"""Passwort-Hashes im Format von Werkzeug.

Damit Weboberfläche (Rust) und Desktop-Oberfläche (Python) dieselbe
Datei data/users.json nutzen können, wird genau das Format von
werkzeug.security erzeugt und geprüft:

* scrypt:n:r:p$salz$hex – Standard neuerer Werkzeug-Fassungen,
* pbkdf2:sha256:runden$salz$hex – Format älterer Fassungen.

Der Vergleich der Hashes läuft zeitunabhängig (hmac.compare_digest).
"""

import hashlib
import hmac
import secrets
import string

#: Länge des abgeleiteten Schlüssels bei scrypt (wie bei Werkzeug).
SCRYPT_LAENGE = 64
#: Standardparameter von Werkzeug für scrypt.
SCRYPT_N = 32768
SCRYPT_R = 8
SCRYPT_P = 1
#: Länge des Salzes in Zeichen (wie bei Werkzeug).
SALZ_LAENGE = 16
#: Zeichenvorrat des Salzes (wie bei Werkzeug: Buchstaben und Ziffern).
SALZ_ZEICHEN = string.ascii_letters + string.digits


def _salz():
    """Erzeugt ein zufälliges Salz."""
    return "".join(secrets.choice(SALZ_ZEICHEN) for _ in range(SALZ_LAENGE))


def _scrypt_hash(passwort, salz, n, r, p):
    """Berechnet den scrypt-Hash eines Passworts."""
    try:
        dk = hashlib.scrypt(
            passwort.encode("utf-8"),
            salt=salz.encode("utf-8"),
            n=n, r=r, p=p,
            dklen=SCRYPT_LAENGE,
            maxmem=132 * 1024 * 1024,
        )
    except (ValueError, TypeError):
        return None
    return dk.hex()


def _pbkdf2_hash(passwort, salz, runden):
    """Berechnet den pbkdf2-Hash eines Passworts (SHA-256)."""
    dk = hashlib.pbkdf2_hmac(
        "sha256",
        passwort.encode("utf-8"),
        salz.encode("utf-8"),
        runden,
        dklen=32,
    )
    return dk.hex()


def erzeuge_hash(passwort):
    """Erzeugt einen Passwort-Hash im Werkzeug-Format (scrypt)."""
    salz = _salz()
    hash_wert = _scrypt_hash(passwort, salz, SCRYPT_N, SCRYPT_R, SCRYPT_P)
    if hash_wert is not None:
        return f"scrypt:{SCRYPT_N}:{SCRYPT_R}:{SCRYPT_P}${salz}${hash_wert}"
    # Sollte nicht vorkommen; dann gilt das ältere, ebenfalls
    # unterstützte Verfahren.
    runden = 1_000_000
    return f"pbkdf2:sha256:{runden}${salz}${_pbkdf2_hash(passwort, salz, runden)}"


def pruefe(hash_wert, passwort):
    """Prüft ein Passwort gegen einen gespeicherten Hash.

    Unbekannte Verfahren und defekte Hashes ergeben False.
    """
    teile = hash_wert.split("$", 2)
    if len(teile) < 3:
        return False
    verfahren, salz, erwartet = teile
    angaben = verfahren.split(":")
    berechnet = None
    if len(angaben) == 4 and angaben[0] == "scrypt":
        try:
            n = int(angaben[1])
            r = int(angaben[2])
            p = int(angaben[3])
        except ValueError:
            return False
        berechnet = _scrypt_hash(passwort, salz, n, r, p)
        if berechnet is None:
            return False
    elif len(angaben) == 3 and angaben[0] == "pbkdf2" and angaben[1] == "sha256":
        try:
            runden = int(angaben[2])
        except ValueError:
            return False
        berechnet = _pbkdf2_hash(passwort, salz, runden)
    elif len(angaben) == 2 and angaben[0] == "pbkdf2" and angaben[1] == "sha256":
        berechnet = _pbkdf2_hash(passwort, salz, 260_000)
    else:
        return False
    return hmac.compare_digest(berechnet, erwartet)


# --- Werkzeug-kompatible API-Namen ---

def hash_passwort(passwort):
    """Erzeugt einen Passwort-Hash im Werkzeug-Format."""
    return erzeuge_hash(passwort)


def pruefe_passwort(passwort, gehasht):
    """Prüft ein Passwort gegen einen gespeicherten Hash."""
    return pruefe(gehasht, passwort)


def hash_benutzer_passwort(passwort, algo="auto"):
    """Erzeugt einen Passwort-Hash im Werkzeug-Format.

    algo kann 'auto', 'scrypt' oder 'pbkdf2' sein.
    """
    if algo in ("auto", "scrypt"):
        return erzeuge_hash(passwort)
    elif algo == "pbkdf2":
        salz = _salz()
        runden = 1_000_000
        return f"pbkdf2:sha256:{runden}${salz}${_pbkdf2_hash(passwort, salz, runden)}"
    return erzeuge_hash(passwort)


def pruefe_benutzer_passwort(passwort, gehasht):
    """Prüft ein Passwort gegen einen gespeicherten Hash."""
    return pruefe(gehasht, passwort)
